packages/eslint-plugin/src/rules/enum-utils/shared.ts

```diff
@@ -33,8 +50,52 @@ export function getEnumTypes(
   typeChecker: ts.TypeChecker,
   type: ts.Type,
 ): ts.Type[] {
-  return tsutils
-    .unionTypeParts(type)
-    .filter(subType => isTypeFlagSet(subType, ts.TypeFlags.EnumLiteral))
-    .map(type => getBaseEnumType(typeChecker, type));
+  return getEnumLiterals(type).map(type => getBaseEnumType(typeChecker, type));
+}
+
+/**
+ * Returns the enum key that matches the given literal node, or null if none
+ * match. For example:
+ * ```ts
+ * enum Fruit {
+ *   Apple = 'apple',
+ *   Banana = 'banana',
+ * }
+ *
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'apple') --> 'Fruit.Apple'
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'banana') --> 'Fruit.Banana'
+ * getEnumKeyForLiteral([Fruit.Apple, Fruit.Banana], 'cherry') --> null
+ * ```
+ */
+export function getEnumKeyForLiteral(
+  enumLiterals: ts.LiteralType[],
+  literal: unknown,
+): string | null {
+  for (const enumLiteral of enumLiterals) {
+    if (enumLiteral.value === literal) {
+      const { symbol } = enumLiteral;
+
+      const memberDeclaration = symbol.valueDeclaration as ts.EnumMember;
+      const enumDeclaration = memberDeclaration.parent;
+
+      const memberNameIdentifier = memberDeclaration.name;
+      const enumName = enumDeclaration.name.text;
+
+      switch (memberNameIdentifier.kind) {
+        case ts.SyntaxKind.Identifier:
+          return `${enumName}.${memberNameIdentifier.text}`;
+
+        case ts.SyntaxKind.StringLiteral:
+          return `${enumName}['${memberNameIdentifier.text}']`;
```

Comment:
not sure what to do in this case, should I put single quotes? double quotes? somehow read it from the source?

Reply:
This brings up a couple necessary test cases:

* `'` with a `\'` inside
* `"` with a `\"` inside

There... may be an existing utility for this? I haven't searched for it.

Reply:
Oh, and https://github.com/typescript-eslint/typescript-eslint/issues/7747 implies the `switch-exhaustiveness-check` fixer for enums might be useful.

Reply:
Seems like it also has some bugs. Enums are awful.
Filed #7768

Reply:
Love it. Let's not introduce new bugs here, and anything existing can be tackled there.

Reply:
@JoshuaKGoldberg 

[I hate enums](https://github.com/typescript-eslint/typescript-eslint/pull/7691/commits/cf560404a626abebff0dc29fcb19d8c21c4cfda4)

anything I've missed?

Reply:
Ah found this thread - yeah we can tackle other issues in #7768. I think you got all the introduced ones!